core/arch/arm/kernel/secure_partition.c

```diff
@@ -1,50 +1,255 @@
+	ctx->uctx.stack_ptr -= ROUNDUP(sizeof(*info), STACK_ALIGNMENT);
+
+	ctx->sp_regs.x[0]  = (uintptr_t)ctx->uctx.stack_ptr;
+	info = (struct sp_ffa_init_info *)ctx->uctx.stack_ptr;
+
+	info->magic = 0;
+	info->count = 0;
 }
 
-REGISTER_SP_STORE(2) = {
-	.description = "SP store",
-	.open = secure_partition_open,
-	.get_size = emb_ts_get_size,
-	.get_tag = emb_ts_get_tag,
-	.read = emb_ts_read,
-	.close = emb_ts_close,
-};
+static uint32_t new_session_id(struct sp_sessions_head *open_sessions)
+{
+	struct sp_session *last = NULL;
+	uint32_t id = MIN_ENDPOINT_ID + 1;
+
+	last = TAILQ_LAST(open_sessions, sp_sessions_head);
+	if (last) {
+		/* This value is less likely to be already used */
+		id = last->endpoint_id + 1;
+	}
+
+	return id;
+}
+
+static TEE_Result sp_create_ctx(const TEE_UUID *uuid, struct sp_session *s)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct sp_ctx *utc = NULL;
+
+	/* Register context */
+	utc = calloc(1, sizeof(struct sp_ctx));
+	if (!utc)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	utc->uctx.ts_ctx = &utc->ts_ctx;
+	utc->open_session = s;
+
+	s->ts_sess.ctx = &utc->ts_ctx;
+
+	utc->ts_ctx.uuid = *uuid;
+	res = vm_info_init(&utc->uctx);
+	if (res)
+		goto err;
+
+	set_sp_ctx_ops(&utc->ts_ctx);
+
+	if (res)
+		goto err;
+
+	vm_set_ctx(NULL);
+	return TEE_SUCCESS;
+
+err:
+	s->ts_sess.ctx = NULL;
+	vm_set_ctx(NULL);
+	pgt_flush_ctx(&utc->ts_ctx);
+
+	tee_pager_rem_um_areas(&utc->uctx);
+	vm_info_final(&utc->uctx);
+	free(utc);
 
-static TEE_Result secure_partition_init(void)
+	return res;
+}
+
+static TEE_Result sp_create_session(struct sp_sessions_head *open_sessions,
+				    const TEE_UUID *uuid,
+				    struct sp_session **sess)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct sp_session *s = calloc(1, sizeof(struct sp_session));
+
+	if (!s)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	s->endpoint_id = new_session_id(open_sessions);
+	if (!s->endpoint_id) {
+		res = TEE_ERROR_OVERFLOW;
+		goto out;
+	}
+
+	DMSG("Loading Secure Partition %pUl", (void *)uuid);
+	res = sp_create_ctx(uuid, s);
+
+out:
+	if (res == TEE_SUCCESS) {
+		TAILQ_INSERT_TAIL(open_sessions, s, link);
+		*sess = s;
+	} else {
+		free(s);
+	}
+	return res;
+}
+
+static TEE_Result sp_init_set_registers(struct sp_ctx *ctx)
 {
+	struct thread_ctx_regs *sp_regs = &ctx->sp_regs;
+	uint64_t x0 = sp_regs->x[0];
+
+	/* Clear all addresses except x0, x0 contains the info address. */
+	memset(sp_regs, 0, sizeof(*sp_regs));
+	sp_regs->sp = ctx->uctx.stack_ptr;
+	sp_regs->x[0] = x0;
+	sp_regs->pc = ctx->uctx.entry_func;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sp_open_session(struct sp_session **sess,
+				  struct sp_sessions_head *open_sessions,
+				  const TEE_UUID *uuid)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct sp_session *s = NULL;
+	struct sp_ctx *ctx = NULL;
+
+	if (!find_secure_partition(uuid))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	res = sp_create_session(open_sessions, uuid, &s);
+	if (res != TEE_SUCCESS) {
+		DMSG("init sp session failed 0x%x", res);
```

Comment:
`DMSG("Init SP session failed %#"PRIx32", res);`
It's not obvious but `TEE_Result` by GP TEE specification so we must use `PRIx32` from [**inttypes.h**](https://github.com/OP-TEE/optee_os/blob/3.12.0/lib/libutils/isoc/include/inttypes.h#L44).